File: vw_py/Pipeline/Pipeline.py
import bindings_vw_py as bindings
from weakref import finalize

class Pipeline:
    def __init__(self, device, render_pass, pipeline_layout, pipeline):
        self.device = device
        self.render_pass = render_pass
        self.pipeline_layout = pipeline_layout
        self.pipeline = pipeline
        finalize(self, bindings.vw_destroy_pipeline, pipeline)

class GraphicsPipelineBuilder:
    def __init__(self, device, render_pass):
        self.device = device
        self.render_pass = render_pass
        self.viewport = None
        self.scissor = None
        self.pipeline_layout = None
        self.number_color_attachment = 0
        self.shader_and_stage_vector = bindings.StageAndShaderVector()
    
    def add_shader(self, stage, module):
        new = bindings.VwStageAndShader()
        new.module = module.shader_module
        new.stage = stage
        self.shader_and_stage_vector.push_back(new)
        return self
    
    def with_fixed_viewport(self, width, height):
        self.viewport = (width, height)
        return self
    
    def with_fixed_scissor(self, width, height):
        self.scissor = (width, height)
        return self
    
    def with_pipeline_layout(self, pipeline_layout):
        self.pipeline_layout = pipeline_layout
        return self
    
    def add_color_attachment(self):
        self.number_color_attachment += 1
        return self
    
    def build(self):
        arguments = bindings.VwGraphicsPipelineCreateArguments()
        arguments.device = self.device.device
        arguments.renderPass = self.render_pass.render_pass
        arguments.number_color_attachment = self.number_color_attachment
        
        if self.pipeline_layout is not None:
            arguments.layout = self.pipeline_layout.pipeline_layout
        
        if self.scissor is not None:
            arguments.withScissor = True
            arguments.widthScissor = self.scissor[0]
            arguments.heightScissor = self.scissor[1]
        
        if self.viewport is not None:
            arguments.withViewport = True
            arguments.widthViewport = self.viewport[0]
            arguments.heightViewport = self.viewport[1]

        arguments.stageAndShaders = bindings.stage_and_shader_vector_data(self.shader_and_stage_vector)
        arguments.size = len(self.shader_and_stage_vector)
        pipeline = bindings.vw_create_graphics_pipeline(arguments)
        return Pipeline(self.device, self.render_pass, self.pipeline_layout, pipeline)

File: vw_rust/src/pipeline/pipeline.rs
use std::ptr;

use super::pipeline_layout::PipelineLayout;
use super::shaders::ShaderModule;
use crate::render_pass::render_pass::RenderPass;
use crate::sys::bindings::{
    self, vw_Pipeline, VkPipeline, VwShaderStageFlagBits, VwGraphicsPipelineCreateArguments,
    VwStageAndShader,
};
use crate::vulkan::device::Device;

pub struct Pipeline<'a> {
    _device: &'a Device<'a>,
    _render_pass: &'a RenderPass<'a>,
    ptr: *mut vw_Pipeline,
}

pub struct GraphicsPipelineBuilder<'a> {
    device: &'a Device<'a>,
    render_pass: &'a RenderPass<'a>,
    stage_and_shaders: Vec<(VwShaderStageFlagBits, ShaderModule<'a>)>,

    viewport: Option<(i32, i32)>,
    scissor: Option<(i32, i32)>,

    pipeline_layout: Option<&'a PipelineLayout<'a>>,
    number_color_attachment: i32,
}

impl<'a> GraphicsPipelineBuilder<'a> {
    pub fn new(
        device: &'a Device<'a>,
        render_pass: &'a RenderPass<'a>,
    ) -> GraphicsPipelineBuilder<'a> {
        GraphicsPipelineBuilder {
            device: device,
            render_pass: render_pass,
            stage_and_shaders: vec![],
            viewport: None,
            scissor: None,
            pipeline_layout: None,
            number_color_attachment: 0,
        }
    }

    pub fn add_shader(
        mut self,
        stage: VwShaderStageFlagBits,
        shader: ShaderModule<'a>,
    ) -> GraphicsPipelineBuilder<'a> {
        self.stage_and_shaders.push((stage, shader));
        self
    }

    pub fn with_fixed_scissor(mut self, size: (i32, i32)) -> GraphicsPipelineBuilder<'a> {
        self.scissor = Some(size);
        self
    }

    pub fn with_fixed_viewport(mut self, size: (i32, i32)) -> GraphicsPipelineBuilder<'a> {
        self.viewport = Some(size);
        self
    }

    pub fn with_pipeline_layout(
        mut self,
        pipeline_layout: &'a PipelineLayout<'a>,
    ) -> GraphicsPipelineBuilder<'a> {
        self.pipeline_layout = Some(pipeline_layout);
        self
    }

    pub fn add_color_attachment(mut self) -> GraphicsPipelineBuilder<'a> {
        self.number_color_attachment += 1;
        self
    }

    pub fn build(self) -> Pipeline<'a> {
        let stage_and_shaders: Vec<_> = self
            .stage_and_shaders
            .iter()
            .map(|(stage, shader)| VwStageAndShader {
                stage: *stage,
                module: shader.as_mut_ptr(),
            })
            .collect();

        let arguments = VwGraphicsPipelineCreateArguments {
            device: self.device.as_ptr(),
            renderPass: self.render_pass.as_ptr(),
            stageAndShaders: stage_and_shaders.as_ptr(),
            size: stage_and_shaders.len() as i32,
            withViewport: self.viewport.is_some(),
            withScissor: self.scissor.is_some(),
            widthViewport: self.viewport.unwrap_or((0, 0)).0,
            heightViewport: self.viewport.unwrap_or((0, 0)).1,
            widthScissor: self.scissor.unwrap_or((0, 0)).0,
            heightScissor: self.scissor.unwrap_or((0, 0)).1,
            layout: match self.pipeline_layout {
                Some(x) => x.as_ptr(),
                _ => ptr::null(),
            },
            number_color_attachment: self.number_color_attachment,
        };

        unsafe {
            Pipeline {
                _device: self.device,
                _render_pass: self.render_pass,
                ptr: bindings::vw_create_graphics_pipeline(&arguments),
            }
        }
    }
}

impl<'a> Pipeline<'a> {
    pub fn as_ptr(&self) -> *const vw_Pipeline {
        self.ptr
    }

    pub fn handle(&self) -> VkPipeline {
        unsafe { bindings::vw_pipeline_handle(self.ptr) }
    }
}

impl<'a> Drop for Pipeline<'a> {
    fn drop(&mut self) {
        unsafe {
            bindings::vw_destroy_pipeline(self.ptr);
        }
    }
}
